### ISSUE TITLE
is there any method to fetch picklist values of fields in object

### ISSUE THREAD

**Issue by ravimadan86** [omniphx/forrest#246]
I am using forrest to fetch data from salesforce and import into my mysql database and vice versa.

I am facing an issue that few fields are drop downs and their dropdown values are set in SF. 

I am looking for a way to fetch the dropdown values so that i can create a select box in my web application to modify field and push back to Salesforce via forrest update method.

Thanks...

**Comment by omniphx** [comment]
Take a look at the Tooling API's Field Definition:https://developer.salesforce.com/docs/atlas.en-us.api_tooling.meta/api_tooling/tooling_api_objects_fielddefinition.htm

One caveat: RecordTypes also determine which picklist values are available. So some picklist values may not actually be available to certain records

**Comment by ravimadan86** [comment]
i would like to rephrase my queries

Could you please give an example query about how to retrieve the custom field picklist value using SOQL query.

I am not getting any workaround for this problem.

for example i have a pick list field 'Gender' in Opportunity Object which has options Male / Female / Other  and I want to fetch all these available option that are assign to this field using SOQL "SELECT ...." Statement.

I would really appreciate your help on this.

Thanks...!!!

**Comment by omniphx** [comment]
Try `SELECT Id, DeveloperName, TableEnumOrId, Metadata FROM CustomField WHERE TableEnumOrId = 'Opportunity' AND DeveloperName = 'Gender'`

Be aware that when querying for metadata it will return a JSON object and you can only query 1 field at a time.

**Comment by ravimadan86** [comment]
my query is 

`$loss = Forrest::query('SELECT Id, DeveloperName, TableEnumOrId, Metadata FROM CustomField WHERE TableEnumOrId=\'Opportunity\' AND DeveloperName=\'Loss_Reason__c\'');`

AND I'm getting exception 

`[
    {
        "message": "\n TableEnumOrId, Metadata FROM CustomField WHERE TableEnumOrId='Opportunity'\n                              ^\nERROR at Row:1:Column:56\nsObject type 'CustomField' is not supported. If you are attempting to use a custom object, be sure to append the '__c' after the entity name. Please reference your WSDL or the describe call for the appropriate names.",
        "errorCode": "INVALID_TYPE"
    }
]`

**Comment by m-lotze** [comment]
The package https://github.com/roblesterjr04/EloquentSalesForce, which uses this package, has a method to get the picklist values of a field:

```
/**
 * Function provided by @seankndy to get picklist values
 *
 * @param  [type] $object [description]
 * @param  [type] $field  [description]
 * @return [type]         [description]
 */
public function getPicklistValues($object, $field)
{
	Forrest::authenticate();
	$desc = Forrest::sobjects($object . '/describe');

	if (!isset($desc['fields'])) {
				return collect([]);
	}

	foreach ($desc['fields'] as $f) {
		if ($f['name'] == $field) {
			$values = [];
			foreach ($f['picklistValues'] as $p) {
				$values[$p['value']] = $p['label'];
			}
			return collect($values);
		}
	}
	return collect([]);
}
```

**Comment by omniphx** [comment]
@rajjanorkar to clarify there are mulitple ways to retrieve picklist values

1. As @m-lotze pointed out, you can use the describe call and traverse through each field for picklist values. The following methods will get those results:
```php
Forrest::sobjects('Account/describe')
Forrest:describe('Account')
``

2. The Tooling API can be used be used to directly query for the picklist entries:
```php
Forrest::tooling("query?q=SELECT+Id,DeveloperName,TableEnumOrId,Metadata+FROM+CustomField+WHERE+TableEnumOrId='Opportunity'+AND+DeveloperName='Gender'")
```

**Comment by ravimadan86** [comment]
@omniphx the 2nd method worked like charm and is more convenient.

Thanks to all of you for helping me out on this. 

I am hereby closing the ticket as the solutions provided worked for me.

Thanks once again!
